src/core/metrics/exporter/retry_socket.rs

```diff
@@ -0,0 +1,118 @@
+//! A TCP Socket wrapper that reconnects automatically.
+use crate::prelude::{IOError, IOErrorKind, IOResult};
+use std::fmt;
+use std::net::SocketAddr;
+use std::time::Duration;
+use tokio::io::AsyncWriteExt;
+use tokio::net::TcpStream;
+use tokio::sync::oneshot::{channel, error::TryRecvError, Receiver};
+use tokio::time::interval;
+
+const MIN_RECONNECT_DELAY_MS: u64 = 50;
+const MAX_RECONNECT_DELAY_MS: u64 = 10_000;
+
+#[derive(Debug)]
+enum State {
+    Socket(TcpStream),
+    Task(Receiver<TcpStream>),
+}
+
+pub(super) struct RetrySocket {
+    address: SocketAddr,
+    socket: State,
+}
+
+impl fmt::Debug for RetrySocket {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.socket.fmt(f)
+    }
+}
+
+impl RetrySocket {
+    pub(super) async fn new(address: SocketAddr) -> Self {
+        let receiver = Self::spawn_task(address.clone());
+        RetrySocket {
+            address,
+            socket: State::Task(receiver),
+        }
+    }
+
+    fn spawn_task(address: SocketAddr) -> Receiver<TcpStream> {
+        let (tx, rx) = channel();
+        tokio::spawn(async move {
+            let mut delay = MIN_RECONNECT_DELAY_MS;
+            loop {
+                if let Ok(stream) = TcpStream::connect(address).await {
```

Comment:
I think it's better to wrap `connect` into `tokio::time::timeout`. Because now it will just block current task. Or it works differently?

Reply:
Hm, on tiny example:
```rust
use tokio::net::TcpStream;

#[tokio::main]
async fn main() {
    let res = TcpStream::connect("127.0.0.1:2003").await;
    println!("{:?}", res);
}
```
This error occurs immediately:
```
Err(Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })
```
So I think, there is no reason to wrap this in `timeout`